File: medium/198_HouseRobber.java
/**
 * 198. House Robber
 * Medium
 * You are a professional robber planning to rob houses along a street.
 * Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them 
 * is that adjacent houses have security systems connected and it will automatically contact the police 
 * if two adjacent houses were broken into on the same night.
 * Given an integer array nums representing the amount of money of each house, return the maximum amount 
 * of money you can rob tonight without alerting the police.
 * 
 * 
 */

class Solution {
    private HashMap<Integer, Integer> memo = new HashMap<Integer, Integer>();
    private int[] nums;
    
    public int rob(int[] nums) {
        this.nums = nums;
        
        return this.dp(nums.length - 1);
    }
    
    private int dp(int i) {
        if (i == 0) return this.nums[0];
        if (i == 1) return Math.max(this.nums[0], this.nums[1]);
        
        if (!this.memo.containsKey(i)) {
            this.memo.put(i, Math.max(this.dp(i - 1), this.dp(i - 2) + nums[i]));
        }

        return this.memo.get(i);
    }
}

File: medium/198_HouseRobber.rs
// 198. House Robber
// medium
// https://leetcode.com/problems/house-robber

use std::cmp::max;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return nums[0];
        }

        let mut dp_prev = nums[0];
        let mut dp = max(nums[0], nums[1]);
        for i in (2..nums.len()) {
            let dp_tmp = max(dp_prev + nums[i], dp);
            dp_prev = dp;
            dp = dp_tmp;
        }

        dp
    }
}